kernel/src/mem.rs

```diff
@@ -1,202 +1,278 @@
-    /// [`AppId`](crate::callback::AppId) changes for any other
-    /// reason, the iterator will be of zero length.
-    pub fn iter(&self) -> slice::Iter<T> {
-        self.as_ref().iter()
+    fn map_or<F, R>(&self, default: R, fun: F) -> R
+    where
+        F: FnOnce(&[u8]) -> R,
+    {
+        match self.process_id {
+            None => default,
+            Some(pid) => pid.kernel.process_map_or(default, pid, |_| {
+                let slice = unsafe { slice::from_raw_parts(self.ptr, self.len) };
+                fun(slice)
+            }),
+        }
     }
+}
 
-    /// Returns an iterator that allows modifying each value
-    ///
-    /// See
-    /// [`std::slice::iter_mut()`](https://doc.rust-lang.org/std/primitive.slice.html#method.iter_mut).
-    ///
-    /// Internally this uses
-    /// [`AsMut`](struct.AppSlice.html#impl-AsMut<[T]>), hence when
-    /// the app dies, restarts or the
-    /// [`AppId`](crate::callback::AppId) changes for any other
-    /// reason, the iterator will be of zero length.
-    pub fn iter_mut(&mut self) -> slice::IterMut<T> {
-        self.as_mut().iter_mut()
+/// Read-only memory region of a process, shared with the kernel
+pub struct ReadOnlyAppSlice {
+    ptr: *const u8,
+    len: usize,
+    process_id: Option<AppId>,
+}
+
+impl ReadOnlyAppSlice {
+    pub(crate) unsafe fn new(ptr: *const u8, len: usize, process_id: AppId) -> Self {
+        ReadOnlyAppSlice {
+            ptr,
+            len,
+            process_id: Some(process_id),
+        }
     }
 
-    /// Iterate over `chunk_size` elements at a time, starting at the
-    /// beginning of the AppSlice.
-    ///
-    /// See
-    /// [`std::slice::chunks()`](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks).
-    ///
-    /// Internally this uses
-    /// [`AsRef`](struct.AppSlice.html#impl-AsRef<[T]>), hence when
-    /// the app dies, restarts or the
-    /// [`AppId`](crate::callback::AppId) changes for any other
-    /// reason, a [`Chunks`](core::slice::Chunks) iterator of zero length will
-    /// be returned.
-    pub fn chunks(&self, size: usize) -> slice::Chunks<T> {
-        self.as_ref().chunks(size)
-    }
-
-    /// Mutably iterate over `chunk_size` elements at a time, starting at the
-    /// beginning of the AppSlice.
-    ///
-    /// See
-    /// [`std::slice::chunks_mut()`](https://doc.rust-lang.org/std/primitive.slice.html#method.chunks_mut).
-    ///
-    /// Internally this uses
-    /// [`AsMut`](struct.AppSlice.html#impl-AsMut<[T]>), hence when
-    /// the app dies, restarts or the
-    /// [`AppId`](crate::callback::AppId) changes for any other
-    /// reason, a [`ChunksMut`](core::slice::ChunksMut) iterator of zero length will
-    /// be returned.
-    ///
-    /// # Panics
-    ///
-    /// Panics if `chunk_size` is 0.
-    pub fn chunks_mut(&mut self, size: usize) -> slice::ChunksMut<T> {
-        self.as_mut().chunks_mut(size)
+    pub unsafe fn new_external(
```

Comment:
This needs safety comment / documentation (and pub interface documentation)

Reply:
#2464